Use property to define shell link name

infeo opened the pull request:
This PR refactors the Autostart service impl to be more customizable.

The hard coded Cryptomator references are removed, instead the used shell link is named with the JVM property `cryptomator.integrationsWin.autoStartShellLinkName`.

Edit: Fixed property name
Edit2: Use snake case
Edit3: Final property name change

overheadhunter commented on src/main/java/org/cryptomator/windows/autostart/WindowsAutoStart.java:
```diff
@@ -11,46 +11,56 @@
 
 		assert exePath.isPresent();
-		int returnCode = winShellLinks.createShortcut(exePath.get(), absoluteStartupEntryPath.toString(), "Cryptomator");
+		int returnCode = winShellLinks.createShortcut(exePath.get(), absoluteStartupEntryPath.toString(), getShellinkName());
```
```suggestion
		int returnCode = winShellLinks.createShortcut(exePath.get(), absoluteStartupEntryPath.toString(), getShellLinkName());
```

overheadhunter commented on src/main/java/org/cryptomator/windows/autostart/WindowsAutoStart.java:
```diff
@@ -59,15 +69,24 @@ public synchronized void enable() throws ToggleAutoStartFailedException {
+			Files.delete(absoluteStartupEntryPath.get());
+			LOG.debug("Successfully deleted {}.", absoluteStartupEntryPath.get());
+		} catch (NoSuchElementException e) {
+			throw new ToggleAutoStartFailedException("Disabling auto start failed using startup folder: Name of shellink is not defined.");
```
```suggestion
			throw new ToggleAutoStartFailedException("Disabling auto start failed using startup folder: Name of shell link is not defined.");
```

overheadhunter commented on src/main/java/org/cryptomator/windows/autostart/WindowsAutoStart.java:
```diff
@@ -59,15 +69,24 @@ public synchronized void enable() throws ToggleAutoStartFailedException {
 		}
 	}
 
+	private String getShellinkName() throws NoSuchElementException {
```
```suggestion
	private String getShellLinkName() throws NoSuchElementException {
```

overheadhunter commented on src/main/java/org/cryptomator/windows/autostart/WindowsAutoStart.java:
```diff
@@ -11,46 +11,56 @@
 		this.winShellLinks = new WinShellLinks();
 		this.exePath = ProcessHandle.current().info().command();
-		this.absoluteStartupEntryPath = Path.of(System.getProperty("user.home"), RELATIVE_STARTUP_FOLDER_ENTRY).toAbsolutePath();
+		this.absoluteStartupEntryPath = () -> Path.of(System.getProperty("user.home"), RELATIVE_STARTUP_FOLDER, this.getShellinkName() + LNK_FILE_EXTENSION).toAbsolutePath();
```
```suggestion
		this.absoluteStartupEntryPath = () -> Path.of(System.getProperty("user.home"), RELATIVE_STARTUP_FOLDER, this.getShellLinkName() + LNK_FILE_EXTENSION).toAbsolutePath();
```

overheadhunter commented on src/main/java/org/cryptomator/windows/autostart/WindowsAutoStart.java:
```diff
@@ -11,46 +11,56 @@
-	private static final String RELATIVE_STARTUP_FOLDER_ENTRY = "\\AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Startup\\Cryptomator.lnk";
+	private static final String RELATIVE_STARTUP_FOLDER = "AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Startup\\";
+	private static final String LNK_FILE_EXTENSION = ".lnk";
+	private static final String LNK_NAME_PROPERTY = "org.cryptomator.integrations_win.autostart.shellLinkName";
```
```suggestion
	private static final String LNK_NAME_PROPERTY = "org.cryptomator.integrations_win.autostart.shell_link_name";
```

overheadhunter commented on README.md:
```diff
@@ -2,6 +2,11 @@
+## Config
+
+This project uses the following JVM properties:
+* `org.cryptomator.integrations_win.autostart.shellLinkName` - Name of the shell link, which is placed in the Windows startup folder to start application on user login
```
```suggestion
* `org.cryptomator.integrations_win.autostart.shell_link_name` - Name of the shell link, which is placed in the Windows startup folder to start application on user login
```

overheadhunter reviewed (changes requested):
* rename remaining occurances
* use snake_case for the system property itself

overheadhunter commented on src/main/java/org/cryptomator/windows/autostart/WindowsAutoStart.java:
```diff
@@ -59,15 +69,24 @@ public synchronized void enable() throws ToggleAutoStartFailedException {
 		}
 	}
 
+	private String getShellLinkName() throws NoSuchElementException {
```
throwing a `NoSuchElementException` just because splitting a path doesn't work as expected is very rude.

overheadhunter commented on src/main/java/org/cryptomator/windows/autostart/WindowsAutoStart.java:
```diff
@@ -59,15 +69,24 @@ public synchronized void enable() throws ToggleAutoStartFailedException {
 
+	private String getShellLinkName() throws NoSuchElementException {
+		var name = System.getProperty(LNK_NAME_PROPERTY);
+		return Objects.requireNonNullElseGet(name, //
+				() -> exePath.map(s -> s.substring(s.lastIndexOf('\\') + 1, s.lastIndexOf('.'))).get() //
+		);
```
Too much crammed into a single line. Less LoC doesn't mean more readable code.

overheadhunter reviewed (changes requested):
I'm wondering why changing an immutable value to something that the code suggests can change over time?

infeo commented:
> I'm wondering why changing an immutable value to something that the code suggests can change over time?

We are using the [`ProcessHandle::info`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/ProcessHandle.Info.html) API, which only return optionals. In case the system property is not set, the startup entry path depends on the presence of the exe path. If it is not present, accessing it throws a `NoSuchElementException`.

To be still able to initalize the object, instead of directly setting the startup entry path, a supplier is used computing it on demand.

infeo commented:
I'll see if i can use an optional instead.

infeo commented:
I added an additional variable and changed the type to Optional<...>. 
Additionally, the behaviour changed: Once an instance is created, the behaviour of this object won't change anymore.
